LPS-139865 Fix PagintationBarTag server side rendering

julien opened the pull request:
The PagintationBarTag wasn't rendering exactly the same html
as it's corresponsing Clay component, this change fixes that.

**Test plan**:
- Fetch this branch
- Deploy the `frontend-taglib-clay` OSGi module
- Assert that the PaginationBarTag renders the same html
  as the Clay component

liferay-continuous-integration commented:
Please only forward necessary changes to Brian Chan during stabilization. Nonurgent changes should wait until the ongoing DXP 7.4 GA1 and Portal 7.4 GA4 release has been completed. For more details on the release timeline and status, see <a href="https://liferay.slack.com/archives/C03BTCQBQ">product-delivery</a>.

liferay-continuous-integration commented:
CI is automatically triggering the following test suites:
- &nbsp;&nbsp;&nbsp;&nbsp;ci:test:**relevant**
- &nbsp;&nbsp;&nbsp;&nbsp;ci:test:**sf**

liferay-continuous-integration commented:
<h1>Build started.</h1><p>Jenkins is currently running tests in the <code>sf</code> test suite.</p><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a></p><h4>Job Summary:</h4><p>Job Link: <a href="https://test-1-30.liferay.com/job/test-portal-source-format/557/">test-portal-source-format</a></p><p><b>For more details click <a href="https://test-1-30.liferay.com//userContent/jobs/test-portal-source-format/builds/557/jenkins-report.html">here</a>.</b></p>

liferay-continuous-integration commented:
<h1>Build started.</h1><p>Jenkins is currently running tests in the <code>relevant</code> test suite.</p><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a></p><h4>Job Summary:</h4><p>Job Link: <a href="https://test-1-26.liferay.com/job/test-portal-acceptance-pullrequest(master)/372/">test-portal-acceptance-pullrequest(master)</a></p><p><b>For more details click <a href="https://test-1-26.liferay.com//userContent/jobs/test-portal-acceptance-pullrequest(master)/builds/372/jenkins-report.html">here</a>.</b></p>

liferay-continuous-integration commented:
<html><h3>:heavy_check_mark: ci:test:sf - 1 out of 1 jobs passed in 2 minutes</h3><details><summary>Click here for more details.</summary><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a><br/>Branch GIT ID: <a href="https://github.com/liferay/liferay-portal/commit/596776d659b936539c7dfed23e61b35bc51a8bb9">596776d659b936539c7dfed23e61b35bc51a8bb9</a></p><h4>Sender Branch:</h4><p>Branch Name: <a href="https://github.com/julien/liferay-portal/tree/LPS-139865">LPS-139865</a><br/>Branch GIT ID: <a href="https://github.com/julien/liferay-portal/commit/950ec6aa9209da03eda33c729d7c87330aac38b4">950ec6aa9209da03eda33c729d7c87330aac38b4</a></p>1 out of 1jobs PASSED<details><summary><strong>1 Successful Jobs:</strong></summary><ul><li><a href="https://test-1-30.liferay.com/job/test-portal-source-format/557/">test-portal-source-format</a></li></ul></details><h5>For more details click <a href="https://test-1-30.liferay.com/userContent/jobs/test-portal-source-format/builds/557/jenkins-report.html">here</a>.</h5></details></html>

julien commented:
Hey @javierdearcos,  

@nhpatt told me I could ask to help me with the review.

Just so you know, this pull request had already been reviewed [here](https://github.com/liferay-frontend/liferay-portal/pull/1501#issuecomment-928496799),
and then manually forwarded [here](https://github.com/brianchandotcom/liferay-portal/pull/107900).

The only difference is that I've renamed some variables according to Brian's 
comment [here](https://github.com/brianchandotcom/liferay-portal/pull/107900#issuecomment-929513707)

> lastIndex, rightBufferStart, leftBufferEnd -> these all seem to be indexes to me.

That's true so I've renamed 
`rightBufferStart` to `rightIndex`
`leftBufferEnd` to `leftIndex`

> itemsToRender, rightBuffer, leftBuffer -> they seem to be the same thing but all have different naming schemes.

These are the same things, and they are the items we want to render, 
So I've renamed:
`itemsToRender` to `centerItems`
`leftBuffer` to `leftItems`
`rightBuffer` to `rightItems`

Let me know if you need any more information. Thanks!

liferay-continuous-integration commented:
<div><div><strong>Jenkins Build:</strong><span></span><a href="https://test-1-30.liferay.com/job/test-portal-source-format/557/">test-portal-source-format#557</a><br/></div><div><strong>Jenkins Report:</strong><span></span><a href="https://test-1-30.liferay.com/userContent/jobs/test-portal-source-format/builds/557/jenkins-report.html">jenkins-report.html</a><br/></div><div><strong>Jenkins Suite:</strong><span></span>sf<br/></div><div><strong>Pull Request:</strong><span></span><a href="https://github.com/liferay-frontend/liferay-portal/pull/1512">liferay-frontend#1512</a><br/></div><div><strong>Testray Routine:</strong><span></span><a href="https://testray.liferay.com/home/-/testray/builds?testrayRoutineId=1306753894">EE Pull Request</a><br/></div><div><strong>Testray Build:</strong><span></span><a href="https://testray.liferay.com/home/-/testray/runs?testrayBuildId=1565078897">[master] ci:test:sf - julien &gt; liferay-frontend - PR#1512 - 2021-09-29[06:37:06]</a><br/></div><div><strong>Testray Importer:</strong><span></span><a href="https://test-1-30.liferay.com/job/publish-testray-report/1819/">publish-testray-report#1819</a><br/></div></div>

liferay-continuous-integration commented:
<html><h3>:heavy_check_mark: ci:test:stable - 10 out of 10 jobs passed</h3><h3>:heavy_check_mark: ci:test:relevant - 24 out of 24 jobs passed in 2 hours 12 minutes</h3><details><summary>Click here for more details.</summary><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a><br/>Branch GIT ID: <a href="https://github.com/liferay/liferay-portal/commit/596776d659b936539c7dfed23e61b35bc51a8bb9">596776d659b936539c7dfed23e61b35bc51a8bb9</a></p><h4>Upstream Comparison:</h4><p>Branch GIT ID: <a href="https://github.com/liferay/liferay-portal/commit/596776d659b936539c7dfed23e61b35bc51a8bb9">596776d659b936539c7dfed23e61b35bc51a8bb9</a><br/>Jenkins Build URL: <a href="https://test-1-1.liferay.com/job/test-portal-acceptance-upstream-dxp(master)//2364">Acceptance Upstream DXP (master) #2364</a></p><details><summary><strong>ci:test:stable - 10 out of 10 jobs PASSED</strong></summary><details><summary><strong>10 Successful Jobs:</strong></summary><ul><li><a href="https://test-1-8.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/442777/">test-portal-acceptance-pullrequest-batch(master)/functional-smoke-tomcat90-mysql57-jdk8_stable/0</a></li><li><a href="https://test-1-25.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/6186/">test-portal-acceptance-pullrequest-batch(master)/functional-upgrade-tomcat90-mysql57-jdk8_stable/0</a></li><li><a href="https://test-1-8.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/442778/">test-portal-acceptance-pullrequest-batch(master)/modules-compile-jdk8/0</a></li><li><a href="https://test-1-22.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/336525/">test-portal-acceptance-pullrequest-batch(master)/modules-integration-mysql57-jdk8_stable/0</a></li><li><a href="https://test-1-6.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/476738/">test-portal-acceptance-pullrequest-batch(master)/modules-unit-jdk8_stable/0</a></li><li><a href="https://test-1-6.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/476741/">test-portal-acceptance-pullrequest-batch(master)/pmd-jdk8/0</a></li><li><a href="https://test-1-6.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/476739/">test-portal-acceptance-pullrequest-batch(master)/portal-license-smoke-jdk8/0</a></li><li><a href="https://test-1-29.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/7032/">test-portal-acceptance-pullrequest-batch(master)/rest-builder-jdk8/0</a></li><li><a href="https://test-1-29.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/7030/">test-portal-acceptance-pullrequest-batch(master)/service-builder-jdk8/0</a></li><li><a href="https://test-1-19.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/471286/">test-portal-acceptance-pullrequest-batch(master)/unit-jdk8_stable/0</a></li></ul></details></details><details><summary><strong>ci:test:relevant - 24 out of 24 jobs PASSED</strong></summary><details><summary><strong>24 Successful Jobs:</strong></summary><ul><li><a href="https://test-1-26.liferay.com/job/test-portal-acceptance-pullrequest(master)/372/">test-portal-acceptance-pullrequest(master)</a></li><li><a href="https://test-1-8.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/442780/">test-portal-acceptance-pullrequest-batch(master)/central-requirements-jdk8/0</a></li><li><a href="https://test-1-8.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/442781/">test-portal-acceptance-pullrequest-batch(master)/empty-temp-dir-jdk8/0</a></li><li><a href="https://test-1-25.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/6185/">test-portal-acceptance-pullrequest-batch(master)/functional-smoke-tomcat90-mysql57-jdk11_open/0</a></li><li><a href="https://test-1-8.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/442777/">test-portal-acceptance-pullrequest-batch(master)/functional-smoke-tomcat90-mysql57-jdk8_stable/0</a></li><li><a href="https://test-1-29.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/7033/">test-portal-acceptance-pullrequest-batch(master)/functional-tomcat90-mysql57-jdk8/0</a></li><li><a href="https://test-1-23.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/333713/">test-portal-acceptance-pullrequest-batch(master)/functional-upgrade-tomcat90-mysql57-jdk8/0</a></li><li><a href="https://test-1-25.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/6186/">test-portal-acceptance-pullrequest-batch(master)/functional-upgrade-tomcat90-mysql57-jdk8_stable/0</a></li><li><a href="https://test-1-19.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/471287/">test-portal-acceptance-pullrequest-batch(master)/lpkg-base-jdk8/0</a></li><li><a href="https://test-1-25.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/6187/">test-portal-acceptance-pullrequest-batch(master)/lpkg-controller-jdk8/0</a></li><li><a href="https://test-1-8.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/442778/">test-portal-acceptance-pullrequest-batch(master)/modules-compile-jdk8/0</a></li><li><a href="https://test-1-25.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/6188/">test-portal-acceptance-pullrequest-batch(master)/modules-integration-mysql57-jdk8/0</a></li><li><a href="https://test-1-22.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/336525/">test-portal-acceptance-pullrequest-batch(master)/modules-integration-mysql57-jdk8_stable/0</a></li><li><a href="https://test-1-8.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/442779/">test-portal-acceptance-pullrequest-batch(master)/modules-semantic-versioning-jdk8/0</a></li><li><a href="https://test-1-31.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/9080/">test-portal-acceptance-pullrequest-batch(master)/modules-unit-jdk8/0</a></li><li><a href="https://test-1-6.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/476738/">test-portal-acceptance-pullrequest-batch(master)/modules-unit-jdk8_stable/0</a></li><li><a href="https://test-1-29.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/7031/">test-portal-acceptance-pullrequest-batch(master)/modules-unit-project-templates-jdk8/0</a></li><li><a href="https://test-1-6.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/476741/">test-portal-acceptance-pullrequest-batch(master)/pmd-jdk8/0</a></li><li><a href="https://test-1-6.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/476739/">test-portal-acceptance-pullrequest-batch(master)/portal-license-smoke-jdk8/0</a></li><li><a href="https://test-1-29.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/7032/">test-portal-acceptance-pullrequest-batch(master)/rest-builder-jdk8/0</a></li><li><a href="https://test-1-6.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/476740/">test-portal-acceptance-pullrequest-batch(master)/semantic-versioning-jdk8/0</a></li><li><a href="https://test-1-29.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/7030/">test-portal-acceptance-pullrequest-batch(master)/service-builder-jdk8/0</a></li><li><a href="https://test-1-29.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/7029/">test-portal-acceptance-pullrequest-batch(master)/unit-jdk8/0</a></li><li><a href="https://test-1-19.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/471286/">test-portal-acceptance-pullrequest-batch(master)/unit-jdk8_stable/0</a></li></ul></details></details><h5>For more details click <a href="https://test-1-26.liferay.com/userContent/jobs/test-portal-acceptance-pullrequest(master)/builds/372/jenkins-report.html">here</a>.</h5></details></html><h5>Test bundle downloads:</h5><ul><li><a href="https://files.liferay.com/private/bundles/test-1-26/jobs/test-portal-acceptance-pullrequest(master)/builds/372/liferay-portal-bundle-tomcat.tar.gz">liferay-portal-bundle-tomcat.tar.gz</a> (mirrors: <a href="http://mirrors.dlc.liferay.com/files.liferay.com/private/bundles/test-1-26/jobs/test-portal-acceptance-pullrequest(master)/builds/372/liferay-portal-bundle-tomcat.tar.gz">dlc</a>, <a href="http://mirrors.lax.liferay.com/files.liferay.com/private/bundles/test-1-26/jobs/test-portal-acceptance-pullrequest(master)/builds/372/liferay-portal-bundle-tomcat.tar.gz">lax</a>)</li></ul>

liferay-continuous-integration commented:
<div><div><strong>Jenkins Build:</strong><span></span><a href="https://test-1-26.liferay.com/job/test-portal-acceptance-pullrequest(master)/372/">test-portal-acceptance-pullrequest(master)#372</a><br/></div><div><strong>Jenkins Report:</strong><span></span><a href="https://test-1-26.liferay.com/userContent/jobs/test-portal-acceptance-pullrequest(master)/builds/372/jenkins-report.html">jenkins-report.html</a><br/></div><div><strong>Jenkins Suite:</strong><span></span>relevant<br/></div><div><strong>Pull Request:</strong><span></span><a href="https://github.com/liferay-frontend/liferay-portal/pull/1512">liferay-frontend#1512</a><br/></div><div><strong>Testray Routine:</strong><span></span><a href="https://testray.liferay.com/home/-/testray/builds?testrayRoutineId=1306753894">EE Pull Request</a><br/></div><div><strong>Testray Build:</strong><span></span><a href="https://testray.liferay.com/home/-/testray/runs?testrayBuildId=1565102654">[master] ci:test:relevant - julien &gt; liferay-frontend - PR#1512 - 2021-09-29[06:37:07]</a><br/></div><div><strong>Testray Importer:</strong><span></span><a href="https://test-1-26.liferay.com/job/publish-testray-report/1058/">publish-testray-report#1058</a><br/></div></div>

julien commented:
green on the first try, this has to be a great pull request :smile:

javierdearcos commented on modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/PaginationBarTag.java:
```diff
@@ -265,19 +269,28 @@ protected int processStartTag() throws Exception {
+		if (_ellipsisBuffer != 0) {
+			int activeIndex = _activePage - 1;
+
+			int[] totalItems = new int[totalPages];
```
I'm a little confused by this **totalItems** name because is almost the same name as **_totalItems** field but means different, but to be honest I can't find a perfect candidate. Maybe totalPageIndexes or pageNumberItems? But the other variables (left, center and right) should follow the same pattern to be related.

javierdearcos commented on modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/PaginationBarTag.java:
```diff
@@ -265,19 +269,28 @@ protected int processStartTag() throws Exception {
 
-		int ellipsisCount = _activePage + _ellipsisBuffer;
+		if (_ellipsisBuffer != 0) {
+			int activeIndex = _activePage - 1;
```
It is better to create the variables just before you need to use them, so maybe we can relocate the activeIndex creation to the bottom just before we need to use it

javierdearcos commented on modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/PaginationBarTag.java:
```diff
@@ -288,50 +301,134 @@ protected int processStartTag() throws Exception {
+				Math.min(activeIndex + _ellipsisBuffer + 1, lastIndex));
+
+			for (int itemToRender : centerItems) {
+				jspWriter.write("<li class=\"page-item");
```
Maybe it is a good strategy to create some auxiliary methods to create the page items to add more meaning, clarify the pagination bar creation strategy and avoid duplicating the code.

You can create a private addPageItem method with the following code
```
jspWriter.write("<li class=\"page-item");

if (itemToRender == _activePage) {
    jspWriter.write(" active");
}

if ((_disabledPages != null) &&
	_disabledPages.contains(Integer.valueOf(itemToRender))) {

    jspWriter.write(" disabled");
}

jspWriter.write("\">");

buttonTag = new ButtonTag();

buttonTag.setCssClass("page-link");
buttonTag.setDisplayType("unstyled");
buttonTag.setLabel(String.valueOf(itemToRender));

buttonTag.doTag(pageContext);
```

Another good candidate is to also create a method for addDropdownPageItem.

Even you can extract methods create the pagination with dropdown and another to create the pagination without dropdown, although Brian usually doesn't like to extract methods that are only called once

julien replied:
@javierdearcos as I mentioned in the previous PR, up to now we decided to keep all the "rendering" part in the `processStartTag` method. If you have a look at [ManagementToolbarTag](https://github.com/liferay/liferay-portal/blob/6fa77fbbc8f10e3fb7418bb2b75a7a62de2143de/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/ManagementToolbarTag.java#L682-L1113), you'll see that it's `processStartTag` method is really "big", when I initially submitted the pull request I had added separate method to avoid repeating myself and avoid having such a big method - so I'd be careful with this.

nhpatt replied:
I suggested the same but I think it's a "taglib pattern" to have everything in the same method.

julien replied:
:+1: Thanks

javierdearcos replied:
Ohh, ok. Didn't know about it. Good to know. So let's leave it as it is. As I said, was just recommendations and nothing mandatory or needed

jbalsas replied:
> I suggested the same but I think it's a "taglib pattern" to have everything in the same method.

I don't think there's such a thing... at least not that I'm aware of. I don't see why we wouldn't want to split the logic in `processStartTag` with some helpers to make it more manageable 🤷

liferay-headless replied:
Julien told me that they tried to extract methods in another taglib. But we can always try :)

nhpatt replied:
Julien told me that they tried to extract methods in another taglib. But we can try :D

jbalsas replied:
Can't remember what happened (and if I was around), but you can check one of our oldest tags like [IcontTag](https://github.com/liferay-frontend/liferay-portal/blob/master/util-taglib/src/com/liferay/taglib/aui/IconTag.java#L91) for this.

julien replied:
Thanks @jbalsas that clarified my doubts. 

I've re-submitted a pull request [here](https://github.com/liferay-frontend/liferay-portal/pull/1525) , Brian [told me](https://github.com/brianchandotcom/liferay-portal/pull/108036#issuecomment-931347386) I could bypass CI, but I want to make sure we're all "happy" with the change.

/cc @javierdearcos

javierdearcos commented on modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/PaginationBarTag.java:
```diff
@@ -265,19 +269,28 @@ protected int processStartTag() throws Exception {
 
-		int totalPages = (int)Math.ceil(_totalItems / _activeDelta);
+		int totalPages = (int)Math.ceil(
+			(double)_totalItems / (double)_activeDelta);
```
I think is enough casting one of the operands

julien replied:
You're probably right and I'm not against it, but I was getting different results than in the React component (I didn't try with just "one" cast though). Just out of curiosity, are casts expensive?

nhpatt replied:
They aren't. Doing one should force the other to also be cast, as JS type coercion. But it's like less idiomatic (as it would be not relying in JS type coercion).

julien replied:
Thanks, great explanation

javierdearcos commented on modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/PaginationBarTag.java:
```diff
@@ -288,50 +301,134 @@ protected int processStartTag() throws Exception {
+			}
+		}
+		else {
+			for (int i = 1; i < (totalPages + 1); i++) {
```
I think you can remove the parenthesis, although maybe is the SourceFormatter who is adding them...

julien replied:
I'm not sure, but I think it was an SF thing

javierdearcos reviewed (approved):
I like the new names. I think they add clarity.

My only concern is with the **totalItems** name as I explained in the comment.

I've added some other suggestions in case you feel they make sense. Feel free to ignore them

nhpatt commented:
@javierdearcos can you do the changes you consider when you have time? that way Brian will see that it has been reviewed by a backend guy.

julien commented:
Hey @javierdearcos and @nhpatt thanks for the feedback.

I added a few comments, I'm OK with all (and thankful for all) these suggestions, my only concern is about "extracting" stuff from the `processStartTag` method, not because of any preference, but just because up to now that's how we've dealt with generating markup in tags (I know, I'm repeating myself :smile:)

julien commented:
Thanks again for the help guys, I'm going to forward this, I'll add @javierdearcos in cc just to show that he participated in the review.

julien commented:
ci:forward

liferay-continuous-integration commented:
CI is automatically triggering the following test suites:
- &nbsp;&nbsp;&nbsp;&nbsp;ci:test:**relevant**
- &nbsp;&nbsp;&nbsp;&nbsp;ci:test:**sf**

The pull request will automatically be forwarded to the user `brianchandotcom` if the following test suites pass:
- &nbsp;&nbsp;&nbsp;&nbsp;ci:test:**relevant**
- &nbsp;&nbsp;&nbsp;&nbsp;ci:test:**sf**
- &nbsp;&nbsp;&nbsp;&nbsp;ci:test:**stable**

liferay-continuous-integration commented:
Skipping previously passed test suites:
`ci:test:relevant`
`ci:test:sf`

liferay-continuous-integration commented:
All required test suite(s) passed.
Forwarding pull request to `brianchandotcom`.
[Console](https://test-1-27.liferay.com/job/forward-pullrequest/68//consoleFull)

liferay-continuous-integration commented:
Pull request has been successfully forwarded to https://github.com/brianchandotcom/liferay-portal/pull/108036
[Console](https://test-1-27.liferay.com/job/forward-pullrequest/68//consoleFull)
